### PR TITLE
Improve robustness of return mapping iterations

### PR DESCRIPTION
**Pull request by bwspenc**
ref #6873
- Adds new base class that provides return mapping capability.  This is
  used currently by solid_mechanics models, but it will be straightforward
  to move tensor_mechanics models over to it.
- Add line search and bisection to drastically improve the robustness of
  return mapping iterations.
- Make all models compute residuals in strain increment units for consistency.
- Use the elastic strain magnitude as the reference quantity for relative
  convergence (rather than the initial residual, which can be extremely high)
- Set the default convergence tolerances to much tighter values
- Remove the option to output iterations on failure.  Instead, always re-run
  the model with output turned on if it fails to converge.
- Add _qp to the ConstitutiveModel base class to be more consistent with
  tensor_mechanics, and make it easier to base the tensor_mechanics models on
  this.
- Add a legacy_return_mapping option to give access to the previous return
  mapping procedure and tolerances as needed.  This is intended only to be
  temporarily available for compatibility with existing models.

### CODE REVIEW

**Comment by bwspenc** [comment]
@sapitts @acasagran @dschwen Would one or more of you mind taking a look at this?

This does break BISON due to API incompatibilites, and I have a patch ready.  With the exception of the couple of cases we discussed that have issues unrelated to this, all of the BISON assessment cases run with this fix, and with only default convergence settings, which are far tighter than the current tolerances.

File: modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+
+#include "InputParameters.h"
+#include "ConsoleStream.h"
+
+/**
+ * Base class that provides capability for Newton return mapping
+ * iterations on a single variable
+ */
+
+class NewtonIterationSingleVariable;
```
Comment (dschwen) [inline modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h]: This name is a bit too general for something that requires a _trial stress_ in my opinion.

**Comment by WilkAndy** [comment]
In TensorMechanics i've created MultiParameterPlasticityStressUpdate (MPPSU) that does the return-map using a line search (between lines 283 and 317 currently).  I think it's all pretty standard except that i check for precision loss because it has turned out to be quite important in plasticity models.  I hope that when you do the equivalent of this PR in TensorMechanics you can either use the code in MPPSU or make MPPSU use your code.  I've spoken to @fdkong about whether MOOSE could include its own Newton-Raphson+line-search because we shouldn't really have multiple bits of code doing the same thing.

**Comment by sapitts** [comment]
@WilkAndy I quite like this idea of a common Newton Rhapson + line search.  @jiangwen84 and I  also have another line search method in our crystal plasticity codes which would further strengthen the argument for a common utility like the one you suggested.

File: modules/solid_mechanics/include/materials/ReturnMappingModel.h [hunk 2]
```diff
@@ -8,61 +8,79 @@
 #define RETURNMAPPINGMODEL_H
 
 #include "ConstitutiveModel.h"
+#include "NewtonIterationSingleVariable.h"
 
 /**
- * One or more constitutive models coupled together.
+ * Base class for models that perform return mapping iterations to compute stress.  This
+ * is for a single model.  This class can be used directly for a single model, or combined
+ * with other models.
```
Comment (sapitts) [inline modules/solid_mechanics/include/materials/ReturnMappingModel.h]: How combined with other models?  Will the multiple models through the picard iterations in tensor mechanics work in a similar fashion to how they do now?

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+  return std::numeric_limits<Real>::max();
+}
+
+void
+NewtonIterationSingleVariable::returnMappingSolve(const Real effective_trial_stress,
+                                                  Real & scalar,
+                                                  const ConsoleStream & console)
+{
+  std::stringstream iter_output;
+  std::stringstream * iter_output_ptr = (_output_iteration_info ? &iter_output : nullptr);
```
Comment (sapitts) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: Is stringstream the right "moosey" thing to use here?

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C [hunk 4]
```diff
@@ -0,0 +1,443 @@
...
+    if (_check_range)
+      checkPermissibleRange(scalar,
+                            scalar_increment,
+                            scalar_old,
+                            min_permissible_scalar,
+                            max_permissible_scalar,
+                            iter_output);
+
+    residual = computeResidualInternal(effective_trial_stress, scalar, dt);
+    reference_residual = computeReferenceResidual(effective_trial_stress, scalar);
```
Comment (sapitts) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: Is this reference residual related to the reference residual in Problem that we use in the assessment cases?

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+}
+
+NewtonIterationSingleVariable::NewtonIterationSingleVariable(const InputParameters & parameters)
+  : _legacy_residual(false),
+    _max_its(parameters.get<unsigned int>("max_its")),
+    _output_iteration_info(parameters.get<bool>("output_iteration_info")),
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance")),
+    _line_search(true),
+    _bracket_solution(true),
```
Comment (sapitts) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: What does this variable do?

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C [hunk 6]
```diff
@@ -0,0 +1,443 @@
...
+        // within the bounds if it is
+        if (scalar_old + scalar_increment >= scalar_upper_bound ||
+            scalar_old + scalar_increment <= scalar_lower_bound)
+        {
+          if (scalar_upper_bound != max_permissible_scalar && scalar_lower_bound != 0.0)
+          {
+            Real frac = 0.5;
+            scalar_increment =
+                (1.0 - frac) * scalar_lower_bound + frac * scalar_upper_bound - scalar_old;
+            // scalar_increment = (scalar_upper_bound + scalar_lower_bound) / 2.0 - scalar_old;
```
Comment (sapitts) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: Remove if unneded

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C [hunk 7]
```diff
@@ -0,0 +1,443 @@
...
+#include "Conversion.h"
+
+template <>
+InputParameters
+validParams<NewtonIterationSingleVariable>()
+{
+  InputParameters params = emptyInputParameters();
+
+  // Newton iteration control parameters
+  params.addParam<unsigned int>("max_its", 200, "Maximum number of Newton iterations");
```
Comment (sapitts) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: While we are breaking BISON, could we change the variable name from _max_its to _max_iterations?

File: modules/solid_mechanics/src/materials/PowerLawCreepModel.C [hunk 8]
```diff
@@ -55,33 +54,33 @@ PowerLawCreepModel::computeStressInitialize(unsigned qp,
...
+PowerLawCreepModel::computeDerivative(const Real effectiveTrialStress, const Real scalar)
 {
-  return -3 * _coefficient * _shear_modulus * _n_exponent *
-             std::pow(effectiveTrialStress - 3 * _shear_modulus * scalar, _n_exponent - 1) *
-             _exponential * _expTime -
-         1 / _dt;
+  return -3.0 * _coefficient * _shear_modulus * _n_exponent *
+             std::pow(effectiveTrialStress - 3.0 * _shear_modulus * scalar, _n_exponent - 1.0) *
+             _exponential * _expTime * _dt -
+         1.0;
```
Comment (sapitts) [inline modules/solid_mechanics/src/materials/PowerLawCreepModel.C]: This looks like a clang format residual; can the 1.0 be moved up to the previous line for code readability without making clang format unhappy?

**Review by sapitts — COMMENTED** [review]
Looks good, and I'm excited to get this capability in both solid and tensor mechanics.  Mainly a number of documentation and variable naming comments, but nothing that would prevent the code from being merged.

File: modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h [hunk 9]
```diff
@@ -0,0 +1,214 @@
...
+public:
+  NewtonIterationSingleVariable(const InputParameters & parameters);
+  virtual ~NewtonIterationSingleVariable() {}
+
+protected:
+  /**
+   * Perform the return mapping iterations
+   * @param effective_trial_stress Effective trial stress
+   * @param scalar                 Inelastic strain increment magnitude being solved for
+   * @param console                Console output
```
Comment (bwspenc) [inline modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h]: I believe this is how we're supposed to document the code for Doxygen.  I've personally never been very convinced of the benefits of Doxygen, but I'm pretty sure that's the coding standard.

File: modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h [hunk 10]
```diff
@@ -0,0 +1,214 @@
...
+   */
+  void returnMappingSolve(const Real effective_trial_stress,
+                          Real & scalar,
+                          const ConsoleStream & console);
+
+  /**
+   * Compute the maximum permissible value of the scalar.  For some models, the magnitude
+   * of this may be known.
+   * @param effective_trial_stress Effective trial stress
```
Comment (bwspenc) [inline modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h]: This is exactly why I don't think the Doxygen comments are particularly useful, but we're supposed to document every single parameter, even if they have self-explanatory names.

File: modules/solid_mechanics/src/materials/CLSHPlasticModel.C [hunk 11]
```diff
@@ -58,46 +54,46 @@ CLSHPlasticModel::computeStressInitialize(unsigned qp,
...
   {
-    Real xflow = _c_beta * (effectiveTrialStress - (3. * _shear_modulus * scalar) -
-                            _hardening_variable[qp] - _yield_stress);
+    const Real xflow = _c_beta * (effectiveTrialStress - (3. * _shear_modulus * scalar) -
+                                  _hardening_variable[_qp] - _yield_stress);
     Real xphi = _c_alpha * std::sinh(xflow);
     _xphidp = -3. * _shear_modulus * _c_alpha * _c_beta * std::cosh(xflow);
     _xphir = -_c_alpha * _c_beta * std::cosh(xflow);
-    residual = xphi - scalar / _dt;
+    residual = xphi * _dt - scalar;
```
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/CLSHPlasticModel.C]: Yes

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C (hunk 7 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: There were enough cases in the BISON assessment cases that failed with a lower number that I decided to bump it up.  Those are usually cases that have crazy predictor stresses on the order of 1e100, and they actually converge if you give them 200 iterations.  The old method computed the residual relative to the initial residual, so it could happily accept a solution with a residual on the order of 1e100.

File: modules/solid_mechanics/include/materials/ReturnMappingModel.h [hunk 12]
```diff
@@ -8,61 +8,79 @@
...
 
-  const unsigned int _max_its;
-  const bool _output_iteration_info;
-  const bool _output_iteration_info_on_error;
-  const Real _relative_tolerance;
-  const Real _absolute_tolerance;
   Real _effective_strain_increment;
 
+  /// 3 * shear modulus
+  Real _three_g;
```
Comment (bwspenc) [inline modules/solid_mechanics/include/materials/ReturnMappingModel.h]: Sure, I can make that change.  I think the standard is that we're supposed to Doxygen coment all of these variables, even if they have self-explanatory names, which gives more opportunities for comments that are lies.  Did I mention that I see no value in Doxygen?

File: modules/solid_mechanics/include/materials/ReturnMappingModel.h [hunk 13]
```diff
@@ -8,61 +8,79 @@
...
 
 /**
- * One or more constitutive models coupled together.
+ * Base class for models that perform return mapping iterations to compute stress.  This
+ * is for a single model.  This class can be used directly for a single model, or combined
+ * with other models.
  */
 
-class ReturnMappingModel : public ConstitutiveModel
+class ReturnMappingModel : public ConstitutiveModel, public NewtonIterationSingleVariable
```
Comment (bwspenc) [inline modules/solid_mechanics/include/materials/ReturnMappingModel.h]: We will follow this same pattern.  RadialReturnStressUpdate will inherit from both StressUpdateBase and NewtonIterationSingleVariable (which I have renamed now), and we'll get rid of all of the Newton iteration code in RadialReturnStressUpdate.  This structure will allow us to accomplish exactly what Andy is suggesting.

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C [hunk 14]
```diff
@@ -0,0 +1,443 @@
...
+      "output_iteration_info", false, "Set true to output Newton iteration information");
+  params.addDeprecatedParam<bool>(
+      "output_iteration_info_on_error",
+      false,
+      "Set true to output Newton iteration information when those iterations fail",
+      "This information is always output when iterations fail");
+  params.addParam<Real>(
+      "relative_tolerance", 1e-8, "Relative convergence tolerance for Newton iteration");
+  params.addParam<Real>(
+      "absolute_tolerance", 1e-11, "Absolute convergence tolerance for Newton iteration");
```
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: I agree with your point, but this is a general base class that isn't intended to prescribe the specifics of that.  That's something determined by the models that derive from this.

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C (hunk 5 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: See my Doxygen comment in the header file.

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C [hunk 15]
```diff
@@ -0,0 +1,443 @@
...
+    _relative_tolerance(parameters.get<Real>("relative_tolerance")),
+    _absolute_tolerance(parameters.get<Real>("absolute_tolerance")),
+    _line_search(true),
+    _bracket_solution(true),
+    _check_range(false)
+{
+  if (parameters.get<bool>("legacy_return_mapping") == true)
+  {
+    if (!parameters.isParamSetByUser("relative_tolerance"))
+      _relative_tolerance = 1.e-5;
```
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: In general, the default was 1e-5, but there is a BISON model that actually set it looser.  I have retained that behavior in my BISON code that accompanies this.

File: modules/solid_mechanics/src/materials/CLSHPlasticModel.C (hunk 11 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/CLSHPlasticModel.C]: Yes

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C (hunk 7 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: I'd be in favor of deprecating max_its.  Maybe we could do that in a separate commit so we don't break more than we have to all at once.

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C (hunk 3 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: I discussed this with @dschwen, and came to the conclusion that there was nothing un-moosey about using a stringstream.  I think it's more natural here.

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C (hunk 4 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: Similar idea.  It's the quantity that we compare the residual to for relative convergence.

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C (hunk 6 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: Done

File: modules/solid_mechanics/src/materials/PowerLawCreepModel.C (hunk 8 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/PowerLawCreepModel.C]: I don't really think so, unfortunately!

File: modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h (hunk 1 above)
Comment (bwspenc) [inline modules/solid_mechanics/include/materials/NewtonIterationSingleVariable.h]: I had the same thought.  I had the idea originally that this would be something used for general single-variable Newton solves, and maybe it's possible to re-working it to be more general in the future.  I propose that for now, we keep this as a return mapping solver (which it appears can already be shared at least 4 different places), and then maybe revisit the idea of making it more general in the future.  I renamed it to SingleVariableReturnMappingSolution to reflect the fact that it is a return mapping solver. Since this is a base class, changing the name won't affect any input files.

**Comment by bwspenc** [comment]
@WilkAndy I wasn't aware of your other version of return mapping iterations.  We should be able to make that derive from this class as well (probably after pulling in a few modifications from your code into it).  I am definitely doing this with the goal of removing duplicated code in mind.  I've put quite a bit of time into making sure that the set of ~80 long-running BISON test cases works with the current code, so I propose we keep this code as-is for now, and then look into the differences between it and your code in the next pass through this.

File: modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C (hunk 14 above)
Comment (bwspenc) [inline modules/solid_mechanics/src/materials/NewtonIterationSingleVariable.C]: Maybe we should redefine this parameter in the derived class for a descriptive string that makes sense for that class?

**Comment by WilkAndy** [comment]
Sounds good to me!  One thing that i haven't implemented is a region of "trust" in the line search.  That is, the NewtonRaphson might send the variables into an unphysical region where we don't want to go, even if the residual is lower there, but the line search (or something else) should be able to prevent this

File: modules/solid_mechanics/include/materials/ReturnMappingModel.h (hunk 2 above)
Comment (bwspenc) [inline modules/solid_mechanics/include/materials/ReturnMappingModel.h]: I revised the comment -- hopefully it makes more sense now!

File: modules/solid_mechanics/include/materials/CombinedCreepPlasticity.h [hunk 16]
```diff
@@ -23,14 +23,12 @@ class CombinedCreepPlasticity : public ConstitutiveModel
...
   /// Compute the stress (sigma += deltaSigma)
   virtual void computeStress(const Elem & current_elem,
-                             unsigned qp,
                              const SymmElasticityTensor & elasticityTensor,
                              const SymmTensor & stress_old,
                              SymmTensor & strain_increment,
                              SymmTensor & stress_new);
 
   virtual bool modifyStrainIncrement(const Elem & current_elem,
-                                     unsigned qp,
```
Comment (bwspenc) [inline modules/solid_mechanics/include/materials/CombinedCreepPlasticity.h]: This still isn't completely foolproof because the model that call it has to set the qp before it calls it, but that only happens in a couple of places in the code, and matches the tensor_mechanics pattern.  In the models themselves, you just always use _qp now, which I think is easier.

**Comment by bwspenc** [comment]
@sapitts Thanks for reviewing this.  I think I addressed all of your issues, and you probably saw that I submitted a corresponding patch to BISON.  If you could review that, I'd appreciate it.  Since getting this pushed through will involve breaking dependent codes, it's probably best if we do this when I get back in the office, probably Monday.  @dschwen, since you're the one who will probably need to merge this, if you could review it so that I can address any issues and have it ready to go Monday, I'd appreciate it.

File: modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h [hunk 17]
```diff
@@ -0,0 +1,191 @@
+/****************************************************************/
+/* MOOSE - Multiphysics Object Oriented Simulation Environment  */
+/*                                                              */
+/*          All contents are licensed under LGPL V2.1           */
+/*             See LICENSE for full restrictions                */
+/****************************************************************/
+#ifndef NEWTONITERATIONSINGLEVARIABLE_H
+#define NEWTONITERATIONSINGLEVARIABLE_H
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h]: the include guard needs to be fixed

File: modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h [hunk 18]
```diff
@@ -0,0 +1,191 @@
...
+
+#include "InputParameters.h"
+#include "ConsoleStream.h"
+
+/**
+ * Base class that provides capability for Newton return mapping
+ * iterations on a single variable
+ */
```
Comment (dschwen) [inline modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h]: can you move that to just above the actual class declaration

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h (hunk 17 above)
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h]: done

File: modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h (hunk 18 above)
Comment (bwspenc) [inline modules/tensor_mechanics/include/materials/SingleVariableReturnMappingSolution.h]: done

**Review by dschwen — APPROVED** [review]
